### PR TITLE
Consolidate compression code

### PR DESCRIPTION
**Pull request by cfergeau**
This pull request should address https://github.com/code-ready/crc/issues/424

### CODE REVIEW

File: pkg/crc/oc/oc_cache.go [hunk 1]
```diff
@@ -43,6 +40,10 @@ func (oc *OcCached) IsCached() bool {
 	return true
 }
 
+func ocFileFilter(filename string) bool {
+	return filepath.Base(filename) == constants.OcBinaryName
```
Comment (gbraad) [inline pkg/crc/oc/oc_cache.go]: Is this returning `true` when the bundle matches the executable name?
if so, why is this called ocFileFilter? why not refering to the fact the binaryname is checked?
And how about `oc.exe` on Windows?

Comment (cfergeau) [reply pkg/crc/oc/oc_cache.go]: > Is this returning true when the bundle matches the executable name?

yes

> if so, why is this called ocFileFilter? why not refering to the fact the binaryname is checked?

Call it `matchOcBinaryName` then?

> And how about oc.exe on Windows?

`pkg/crc/constants/constants_windows.go: OcBinaryName = "oc.exe"`

**Comment by cfergeau** [comment]
retest this please

**Comment by praveenkumar** [comment]
@cfergeau can you resolve the conflict for this and rebase.

**Comment by praveenkumar** [comment]
@cfergeau I think this just need rebase and we will merge it after test in windows.

**Comment by cfergeau** [comment]
retest this please

**Comment by praveenkumar** [comment]
retest this please.

File: pkg/extract/extract.go [hunk 2]
```diff
@@ -34,12 +37,52 @@ func Ungzip(source, target string) error {
...
 
+func Uncompress(tarball, targetDir string) error {
+	logging.Debugf("Uncompressing %s to %s", tarball, targetDir)
+
+	if strings.HasSuffix(tarball, ".zip") {
+		return Unzip(tarball, targetDir)
+	}
+
+	var filereader io.Reader
+	file, err := os.Open(filepath.Clean(tarball))
```
Comment (praveenkumar) [inline pkg/extract/extract.go]: :+1

File: pkg/extract/extract.go [hunk 3]
```diff
@@ -34,12 +37,52 @@ func Ungzip(source, target string) error {
...
+		}
+		defer reader.Close()
+		filereader = io.Reader(reader)
+	} else if strings.HasSuffix(tarball, ".tar") {
+		filereader = file
+	} else {
+		logging.Warnf("Unknown file format when trying to uncompress %s", tarball)
+	}
+
+	return untar(filereader, targetDir)
```
Comment (praveenkumar) [inline pkg/extract/extract.go]: :+1:

File: pkg/crc/machine/bundle/metadata.go [hunk 4]
```diff
@@ -89,6 +91,15 @@ func (bundle *CrcBundleInfo) resolvePath(filename string) string {
 	return filepath.Join(bundle.cachedPath, filename)
 }
 
+func Extract(sourcepath string) (*CrcBundleInfo, error) {
+	err := extract.Uncompress(sourcepath, constants.MachineCacheDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return GetCachedBundleInfo(filepath.Base(sourcepath))
```
Comment (praveenkumar) [inline pkg/crc/machine/bundle/metadata.go]: :+1:

**Comment by praveenkumar** [comment]
I tested this patch on Windows and Mac works as expected.

**Pull request merged**